### PR TITLE
Gameplay upgrade

### PR DESCRIPTION
**Pull request by niemarcin**

### CODE REVIEW

File: shm/Alcohol.cpp [hunk 1]
```diff
@@ -19,13 +19,13 @@ Cargo& Alcohol::operator-=(const size_t amount) {
 }
 
 bool Alcohol::operator==(const Cargo& cargo) const {
-    return getName() == cargo.getName() &&
+    return getName() == cargo.getName();/* &&
```
Comment (J-Cieplinski) [inline shm/Alcohol.cpp]: If you gonna do this then just remove code outright instead of commenting

File: shm/Game.cpp [hunk 2]
```diff
@@ -260,6 +271,8 @@ void Game::makeAction(Action choice) {
 
 void Game::startGame() {
     while (true) {
+        currentPosition_ = map_->getCurrentPosition();
```
Comment (J-Cieplinski) [inline shm/Game.cpp]: There is no need for it to be called on every loop. Updating it after travel would be better

File: shm/Game.hpp [hunk 3]
```diff
@@ -19,6 +19,7 @@ class Game {
     Map* map_{nullptr};
     std::shared_ptr<Player> player_{nullptr};
     Time* time_{nullptr};
+    Island* currentPosition_{nullptr};
```
Comment (J-Cieplinski) [inline shm/Game.hpp]: I've been working on converting raw pointers to smart ones(forgot to push). can this be unique_ptr?

File: shm/Item.cpp [hunk 4]
```diff
@@ -19,13 +19,28 @@ Cargo& Item::operator-=(const size_t amount) {
...
+    case Rarity::common:
+        return "common";
+    case Rarity::rare:
+        return "rare";
+    case Rarity::epic:
+        return "epic";
+    case Rarity::legendary:
+        return "legendary";
+    default:
+        return "";
```
Comment (J-Cieplinski) [inline shm/Item.cpp]: I would go with uppercase

**Review by J-Cieplinski — CHANGES_REQUESTED** [review]

File: shm/Store.cpp [hunk 5]
```diff
@@ -101,17 +101,31 @@ Response Store::buy(const CargoPtr& cargo, size_t amount, const std::shared_ptr<
...
+    double unitPrice{0.0};
+
+    if (!cargoFoundInStore) {
+        unitPrice = playersCargo->getPrice();
+    }
+    else {
+        unitPrice = cargoFoundInStore->getPrice();
+        if ((playersCargo->getUniqueStat() < cargoFoundInStore->getUniqueStat()) 
+                && (typeid(*playersCargo) == typeid(Fruit) || typeid(*playersCargo) == typeid(Alcohol))) {
+            unitPrice *= (double)(cargoFoundInStore->getUniqueStat() + playersCargo->getUniqueStat()) / 2;
```
Comment (iprice73) [inline shm/Store.cpp]: I think it's better to use `static_cast<double>`

File: shm/Store.cpp [hunk 6]
```diff
@@ -101,17 +101,31 @@ Response Store::buy(const CargoPtr& cargo, size_t amount, const std::shared_ptr<
...
+            unitPrice /= cargoFoundInStore->getUniqueStat();
+        }
     }
+    size_t price = (size_t)(unitPrice * amount);
```
Comment (iprice73) [inline shm/Store.cpp]: same here, `static_cast<size_t>`

**Review by iprice73 — CHANGES_REQUESTED** [review]

File: shm/Item.cpp (hunk 4 above)
Comment (niemarcin) [reply shm/Item.cpp]: You mean first letter uppercase or whole string uppercase?

File: shm/Game.hpp (hunk 3 above)
Comment (niemarcin) [reply shm/Game.hpp]: done

**Review by lisie-sprawy — APPROVED** [review]

File: shm/Game.cpp [hunk 7]
```diff
@@ -100,19 +100,20 @@ void Game::travel() {
...
             printPromptCurrentPositionEqualsDestination();
         }
         printCurrentPositionOnMap();
         destination = map_->getIsland(getTravelLocation());
     }
 
     auto distance = map_->getDistanceToIsland(destination);
-    map_->setCurrentPosition(destination);
+    currentPosition_ = std::make_unique<Island>(*destination);
+    map_->setCurrentPosition(currentPosition_.get());
```
Comment (J-Cieplinski) [inline shm/Game.cpp]: Why construct a new pointer from existing object raw pointer to just pass a raw pointer to function? I don't think this code is needed
```suggestion
    currentPosition_ = std::make_unique<Island>(*destination);
    map_->setCurrentPosition(currentPosition_.get());
```

Comment (J-Cieplinski) [reply shm/Game.cpp]: Also, upon leaving this function the smart pointer will destroy itself and we will probably be left with a dangling pointer. So it's unnecessary and dangerous

File: shm/Game.hpp [hunk 8]
```diff
@@ -19,7 +19,7 @@ class Game {
     Map* map_{nullptr};
     std::shared_ptr<Player> player_{nullptr};
     Time* time_{nullptr};
-    Island* currentPosition_{nullptr};
+    std::unique_ptr<Island> currentPosition_{nullptr};
```
Comment (J-Cieplinski) [inline shm/Game.hpp]: I have no idea why it's being done this way. If you want current position to be shared then make it shared pointer not a unique one. Getting a raw pointer and from object of that pointer constructing a new unique pointer defeats the purpose. Just change it to shared pointer and make function which return raw pointers return a shared one instead.

**Review by J-Cieplinski — CHANGES_REQUESTED** [review]

File: shm/Game.hpp (hunk 8 above)
Comment (J-Cieplinski) [reply shm/Game.hpp]: Also, I know that it was my suggestion to convert it to unique_ptr. But at the time I had no idea how it would be used :)

**Pull request merged**